src/utils/Transaction.js

```diff
@@ -180,6 +180,48 @@ const tryToMergeWithLeft = (structs, pos) => {
   }
 }
 
+/**
+ * @param {Array<AbstractStruct>} structs
+ * @param {number} pos
+ */
+const tryToMergeWithLeftNoSplice = (structs, pos) => {
+  const right = structs[pos]
+  if (/** @type any */(right)._wasMerged) {
+    return 0
+  }
+  let offset = 1
+  let left
+  do {
+    left = structs[pos - offset++]
+  } while (/** @type any */(left)._wasMerged)
+  if (left.deleted === right.deleted && left.constructor === right.constructor) {
+    if (left.mergeWith(right)) {
+      /** @type any */(right)._wasMerged = true
+      if (right instanceof Item && right.parentSub !== null && /** @type {AbstractType<any>} */ (right.parent)._map.get(right.parentSub) === right) {
+        /** @type {AbstractType<any>} */ (right.parent)._map.set(right.parentSub, /** @type {Item} */ (left))
+      }
+      return 1
+    }
+  }
+  return 0
+}
+
+/**
+ * @param {Array<AbstractStruct>} structs
+ * @param {number} mergedCount
+ */
+const filterMergedStructs = (structs, mergedCount) => {
```

Comment:
probably better off constructing a new array using array.filter but im not sure if there will be references to the old array dangling so gotta do this ugly mutation